From: =?UTF-8?q?La=C3=A9rcio=20de=20Sousa?= <lbsousajr@gmail.com>
Date: Thu, 7 Nov 2013 09:31:56 -0200
Subject: [PATCH] xfree86: Keep a non-seat0 X server from touching VTs
 (#71258)

diff --git a/hw/xfree86/common/xf86Init.c b/hw/xfree86/common/xf86Init.c
index 788f808..52f1a0a 100644
--- a/hw/xfree86/common/xf86Init.c
+++ b/hw/xfree86/common/xf86Init.c
@@ -564,7 +564,7 @@ InitOutput(ScreenInfo * pScreenInfo, int argc, char **argv)
                     want_hw_access = TRUE;
 
                 if (!(flags & HW_SKIP_CONSOLE))
-                    xorgHWOpenConsole = TRUE;
+                    xorgHWOpenConsole = !ServerIsNotSeat0();
             }
         }
 
